runtime/dynstats.c

```diff
@@ -499,12 +601,151 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
 	RETiRet;
 }
 
+/* not supported in older platforms - disable */
+#ifndef O_CLOEXEC
+#define O_CLOEXEC 0
+#endif
+/*
+ *	Helper function adapted from imfile of the same name - possible candidate for refactor.
+ */
+static rsRetVal ATTR_NONNULL()
+atomicWriteStateFile(const char *fn, const char *content) {
+	DEFiRet;
+	const int fd = open(fn, O_CLOEXEC | O_NOCTTY | O_WRONLY | O_CREAT | O_TRUNC, 0600);
+	if(fd < 0) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: cannot open state file '%s' for "
+			"persisting file state - some data will probably be duplicated "
+			"on next startup", fn);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+	const size_t toWrite = strlen(content);
+	const ssize_t w = write(fd, content, toWrite);
+	if(w != (ssize_t) toWrite) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: partial write to state file '%s' "
+			"this may cause trouble in the future. We will try to delete the "
+			"state file, as this provides most consistent state", fn);
+		unlink(fn);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
```

Comment:
Shouldn't we re-attempt partial writes? (the case where `w < toWrite`)?

Partial write may not be due to an unrecoverable error.

Reply:
I think this depends on whether we think the re-attempt should occur right away, or on the next attempt to persist as specified by the user. Do you think that is sufficient, or we should be more proactive on this?